### ISSUE TITLE
Omnisharp uses wrong MSBuild version

### ISSUE THREAD

**Issue by jwbay** [OmniSharp/omnisharp-vscode#3074]
## Issue Description ##
OmniSharp is selecting the wrong MSBuild location when opening .NET Core 2 applications. The logs tell the story:
```
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 3 MSBuild instance(s)
            1: Visual Studio Build Tools 2019 16.0.28803.452 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin"
            2: Visual Studio Professional 2017 15.8.28010.2041 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin"
            3: StandAlone 15.0 - "C:\Users\x\.vscode\extensions\ms-vscode.csharp-1.19.1\.omnisharp\1.32.19\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: Visual Studio Professional 2017 15.8.28010.2041 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin"
```
Here we see that even though MSBuild 16 is detected via VS Build Tools 2019, it chooses MSBuild 15. Then:
```
[fail]: OmniSharp.MSBuild.ProjectLoader
        Project X is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Project X supports: netcoreapp2.2 (.NETCoreApp,Version=v2.2)
[fail]: OmniSharp.MSBuild.ProjectLoader
        Version 2.2.204 of the .NET Core SDK requires at least version 16.0.0 of MSBuild. The current available version of MSBuild is 15.8.169.51996. Change the .NET Core SDK specified in global.json to an older version that requires the MSBuild version currently available.
[fail]: OmniSharp.MSBuild.ProjectLoader
        The SDK 'Microsoft.NET.Sdk.Web' specified could not be found.
```
We see that Omnisharp is complaining about not having version 16 of MSBuild available. But as we saw above, it's right there 😭. Weirdly, it suggests downgrading the SDK rather than fixing MSBuild. This seems exactly backwards for the majority case.

I tried creating an omnisharp.json like this:
```json
{
    "MsBuild": {
        "MSBuildSdksPath": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\MSBuild\\Current\\Bin"
    }
}
```
But it had no effect. I also tried pointing it to the dotnet sdk folder. That didn't do anything, either.

## Steps to Reproduce ##
* Install VS 2017
* Install VS 2019 build tooling
* Create an ASP NetCore 2 app and open it in VSCode

## Expected Behavior ##
Project is loaded successfully. _Visual Studio installs don't mess with OmniSharp._

## Actual Behavior ##
Project is not loaded. OmniSharp behavior is coupled to installs of VS.

### OmniSharp log ###
I hit this issue in the middle of a project upgrade from DotNetCore 1.0 to 2.0, so there is a fair amount of noise here about broken projects.
<details>
Starting OmniSharp server at 5/18/2019, 12:26:01 PM
    Target: c:\Source\Project\Project.sln

OmniSharp server started.
    Path: C:\Users\user\.vscode\extensions\ms-vscode.csharp-1.19.1\.omnisharp\1.32.19\OmniSharp.exe
    PID: 3860

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on Windows 6.2.9200.0 (x64)
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 3 MSBuild instance(s)
            1: Visual Studio Build Tools 2019 16.0.28803.452 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin"
            2: Visual Studio Professional 2017 15.8.28010.2041 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin"
            3: StandAlone 15.0 - "C:\Users\user\.vscode\extensions\ms-vscode.csharp-1.19.1\.omnisharp\1.32.19\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: Visual Studio Professional 2017 15.8.28010.2041 - "C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\15.0\Bin"
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in 'c:\Source\Project'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Could not find any Cake files
[info]: OmniSharp.WorkspaceInitializer
        Project system 'OmniSharp.DotNet.DotNetProjectSystem' is disabled in the configuration.
[info]: OmniSharp.MSBuild.ProjectSystem
        Detecting projects in 'c:\Source\Project\Project.sln'.
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Source\Project\src\Project.Generation\Project.Generation.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Source\Project\src\Project.Server\Project.Server.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Source\Project\test\Project.Server.IntegrationTests\Project.Server.IntegrationTests.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\Source\Project\test\Project.Server.Tests\Project.Server.Tests.csproj'
[info]: OmniSharp.Script.ScriptProjectSystem
        Detecting CSX files in 'c:\Source\Project'.
[info]: OmniSharp.Script.ScriptProjectSystem
        Could not find any CSX files
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpWorkspaceOptionsProvider
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Source\Project\src\Project.Generation\Project.Generation.csproj
[info]: OmniSharp.WorkspaceInitializer
        Configuration finished.
[info]: OmniSharp.Stdio.Host
        Omnisharp server running using Stdio at location 'c:\Source\Project' on host 16876.
[fail]: OmniSharp.MSBuild.ProjectLoader
        Project Project.Server is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Project Project.Server supports: netcoreapp2.2 (.NETCoreApp,Version=v2.2)
[fail]: OmniSharp.MSBuild.ProjectLoader
        Version 2.2.204 of the .NET Core SDK requires at least version 16.0.0 of MSBuild. The current available version of MSBuild is 15.8.169.51996. Change the .NET Core SDK specified in global.json to an older version that requires the MSBuild version currently available.
[fail]: OmniSharp.MSBuild.ProjectLoader
        The SDK 'Microsoft.NET.Sdk.Web' specified could not be found.
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Source\Project\src\Project.Generation\Project.Generation.csproj'.
c:\Source\Project\src\Project.Generation\Project.Generation.csproj
C:\Source\Project\src\Project.Generation\Project.Generation.csproj(0,0): Error: Project Project.Server is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Project Project.Server supports: netcoreapp2.2 (.NETCoreApp,Version=v2.2)
c:\Source\Project\src\Project.Server\Project.Server.csproj(0,0): Error: Version 2.2.204 of the .NET Core SDK requires at least version 16.0.0 of MSBuild. The current available version of MSBuild is 15.8.169.51996. Change the .NET Core SDK specified in global.json to an older version that requires the MSBuild version currently available.
c:\Source\Project\src\Project.Server\Project.Server.csproj(0,0): Error: The SDK 'Microsoft.NET.Sdk.Web' specified could not be found.

[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Source\Project\src\Project.Server\Project.Server.csproj
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Source\Project\src\Project.Server\Project.Server.csproj'.
c:\Source\Project\src\Project.Server\Project.Server.csproj(1,1)
Microsoft.Build.Exceptions.InvalidProjectFileException: The SDK 'Microsoft.NET.Sdk.Web' specified could not be found.  c:\Source\Project\src\Project.Server\Project.Server.csproj
   at Microsoft.Build.Shared.ProjectErrorUtilities.ThrowInvalidProject(String errorSubCategoryResourceName, IElementLocation elementLocation, String resourceName, Object[] args)
   at Microsoft.Build.Evaluation.Evaluator`4.ExpandAndLoadImportsFromUnescapedImportExpressionConditioned(String directoryOfImportingFile, ProjectImportElement importElement, List`1& projects, SdkResult& sdkResult, Boolean throwOnFileNotExistsError)
   at Microsoft.Build.Evaluation.Evaluator`4.ExpandAndLoadImports(String directoryOfImportingFile, ProjectImportElement importElement, SdkResult& sdkResult)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement)
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
   at Microsoft.Build.Evaluation.Evaluator`4.Evaluate(ILoggingService loggingService, BuildEventContext buildEventContext)
   at Microsoft.Build.Evaluation.Project.Reevaluate(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.Project.ReevaluateIfNecessary(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.Project.Initialize(IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.Project..ctor(String projectFile, IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectCollection projectCollection, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Microsoft.Build.Evaluation.ProjectCollection.LoadProject(String fileName, IDictionary`2 globalProperties, String toolsVersion)
   at OmniSharp.MSBuild.ProjectLoader.EvaluateProjectFileCore(String filePath) in C:\projects\omnisharp-roslyn\src\OmniSharp.MSBuild\ProjectLoader.cs:line 129
   at OmniSharp.MSBuild.ProjectLoader.BuildProject(String filePath) in C:\projects\omnisharp-roslyn\src\OmniSharp.MSBuild\ProjectLoader.cs:line 72
   at OmniSharp.MSBuild.ProjectFile.ProjectFileInfo.Load(String filePath, ProjectIdInfo projectIdInfo, ProjectLoader loader) in C:\projects\omnisharp-roslyn\src\OmniSharp.MSBuild\ProjectFile\ProjectFileInfo.cs:line 94
   at OmniSharp.MSBuild.ProjectManager.<>c__DisplayClass30_0.<LoadProject>b__0() in C:\projects\omnisharp-roslyn\src\OmniSharp.MSBuild\ProjectManager.cs:line 293
   at OmniSharp.MSBuild.ProjectManager.LoadOrReloadProject(String projectFilePath, Func`1 loader) in C:\projects\omnisharp-roslyn\src\OmniSharp.MSBuild\ProjectManager.cs:line 304

[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Source\Project\test\Project.Server.IntegrationTests\Project.Server.IntegrationTests.csproj
[fail]: OmniSharp.MSBuild.ProjectLoader
        Project Project.Server is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Project Project.Server supports: netcoreapp2.2 (.NETCoreApp,Version=v2.2)
[fail]: OmniSharp.MSBuild.ProjectLoader
        Version 2.2.204 of the .NET Core SDK requires at least version 16.0.0 of MSBuild. The current available version of MSBuild is 15.8.169.51996. Change the .NET Core SDK specified in global.json to an older version that requires the MSBuild version currently available.
[fail]: OmniSharp.MSBuild.ProjectLoader
        The SDK 'Microsoft.NET.Sdk.Web' specified could not be found.
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Source\Project\test\Project.Server.IntegrationTests\Project.Server.IntegrationTests.csproj'.
c:\Source\Project\test\Project.Server.IntegrationTests\Project.Server.IntegrationTests.csproj
C:\Source\Project\test\Project.Server.IntegrationTests\Project.Server.IntegrationTests.csproj(0,0): Error: Project Project.Server is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Project Project.Server supports: netcoreapp2.2 (.NETCoreApp,Version=v2.2)
c:\Source\Project\src\Project.Server\Project.Server.csproj(0,0): Error: Version 2.2.204 of the .NET Core SDK requires at least version 16.0.0 of MSBuild. The current available version of MSBuild is 15.8.169.51996. Change the .NET Core SDK specified in global.json to an older version that requires the MSBuild version currently available.
c:\Source\Project\src\Project.Server\Project.Server.csproj(0,0): Error: The SDK 'Microsoft.NET.Sdk.Web' specified could not be found.

[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\Source\Project\test\Project.Server.Tests\Project.Server.Tests.csproj
[fail]: OmniSharp.MSBuild.ProjectLoader
        Project Project.Server is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Project Project.Server supports: netcoreapp2.2 (.NETCoreApp,Version=v2.2)
[fail]: OmniSharp.MSBuild.ProjectLoader
        Version 2.2.204 of the .NET Core SDK requires at least version 16.0.0 of MSBuild. The current available version of MSBuild is 15.8.169.51996. Change the .NET Core SDK specified in global.json to an older version that requires the MSBuild version currently available.
[fail]: OmniSharp.MSBuild.ProjectLoader
        The SDK 'Microsoft.NET.Sdk.Web' specified could not be found.
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'c:\Source\Project\test\Project.Server.Tests\Project.Server.Tests.csproj'.
c:\Source\Project\test\Project.Server.Tests\Project.Server.Tests.csproj
C:\Source\Project\test\Project.Server.Tests\Project.Server.Tests.csproj(0,0): Error: Project Project.Server is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Project Project.Server supports: netcoreapp2.2 (.NETCoreApp,Version=v2.2)
c:\Source\Project\src\Project.Server\Project.Server.csproj(0,0): Error: Version 2.2.204 of the .NET Core SDK requires at least version 16.0.0 of MSBuild. The current available version of MSBuild is 15.8.169.51996. Change the .NET Core SDK specified in global.json to an older version that requires the MSBuild version currently available.
c:\Source\Project\src\Project.Server\Project.Server.csproj(0,0): Error: The SDK 'Microsoft.NET.Sdk.Web' specified could not be found.

[fail]: OmniSharp.MSBuild.ProjectManager
        Attempted to update project that is not loaded: c:\Source\Project\src\Project.Generation\Project.Generation.csproj
[fail]: OmniSharp.MSBuild.ProjectManager
        Attempted to update project that is not loaded: c:\Source\Project\src\Project.Server\Project.Server.csproj
[fail]: OmniSharp.MSBuild.ProjectManager
        Attempted to update project that is not loaded: c:\Source\Project\test\Project.Server.IntegrationTests\Project.Server.IntegrationTests.csproj
[fail]: OmniSharp.MSBuild.ProjectManager
        Attempted to update project that is not loaded: c:\Source\Project\test\Project.Server.Tests\Project.Server.Tests.csproj

</details>

### C# log ###
<details>
Empty
</details>

## Environment information ##

**VSCode version**: 1.34.0
**C# Extension**: 1.19.1

<details><summary>Dotnet Information</summary>
.NET Core SDK (reflecting any global.json):
 Version:   2.2.204
 Commit:    8757db13ec

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.204\

Host (useful for support):
  Version: 2.2.5
  Commit:  0a3c9209c0

.NET Core SDKs installed:
  1.1.0 [C:\Program Files\dotnet\sdk]
  2.2.204 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
</details>

**Comment by franlyreynoso** [comment]
Hi, 
I'm facing the same issue.
I don't know all the effects this issue is causing in vs-code. I can't use "Go To Definition" command and sometimes intellisense simply doesn't work as expected.

Omnisharp log below
----------------------------------------
Starting OmniSharp server at 10/15/2019, 9:55:57 AM
    Target: c:\dev\WebApp1

OmniSharp server started.
    Path: c:\Users\Username\.vscode\extensions\ms-vscode.csharp-1.21.4\.omnisharp\1.34.4\OmniSharp.exe
    PID: 4892

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on Windows 6.2.9200.0 (x64)
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
**_[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 2 MSBuild instance(s)
            1: Visual Studio Community 2019 16.3.29409.12 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"
            2: StandAlone 16.0 - "C:\Users\Username\.vscode\extensions\ms-vscode.csharp-1.21.4\.omnisharp\1.34.4\.msbuild\Current\Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: Visual Studio Community 2019 16.3.29409.12 - "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin"_**
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in 'c:\dev\WebApp1'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Could not find any Cake files
[info]: OmniSharp.MSBuild.ProjectSystem
        No solution files found in 'c:\dev\WebApp1'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'c:\dev\WebApp1\WebApp1.csproj'
[info]: OmniSharp.Script.ScriptProjectSystem
        Detecting CSX files in 'c:\dev\WebApp1'.
[info]: OmniSharp.Script.ScriptProjectSystem
        Could not find any CSX files
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: c:\dev\WebApp1\WebApp1.csproj
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.RenameWorkspaceOptionsProvider, Order: 100
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.EditorConfigWorkspaceOptionsProvider, Order: 200
[info]: OmniSharp.WorkspaceInitializer
        Configuration finished.
[info]: OmniSharp.Stdio.Host
        Omnisharp server running using Stdio at location 'c:\dev\WebApp1' on host 17620.
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file 'c:\dev\WebApp1\WebApp1.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project 'c:\dev\WebApp1\WebApp1.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: WebApp1

**Comment by filipw** [comment]
in fact I will close this issue because this has been resolved - OmniSharp now ships with 16.3 MSBuild and supports .NET Core 3.0 too.

@franlyreynoso it uses the correct one - VS 2019 version of msbuild is preferred over the stand alone one. "go to definition" issue is not a related thing - it's a regression in 1.21.4, see https://github.com/OmniSharp/omnisharp-vscode/issues/3344#issuecomment-541553080 for a workaround

**Comment by franlyreynoso** [comment]
> 
> 
> in fact I will close this issue because this has been resolved - OmniSharp now ships with 16.3 MSBuild and supports .NET Core 3.0 too.
> 
> @franlyreynoso it uses the correct one - VS 2019 version of msbuild is preferred over the stand alone one. "go to definition" issue is not a related thing - it's a regression in 1.21.4, see [#3344 (comment)](https://github.com/OmniSharp/omnisharp-vscode/issues/3344#issuecomment-541553080) for a workaround

Thanks! That fixed it.
